File: Fourth Semester - Design and Programming Environments/C#/ProiectC#/model/Reservation.cs
namespace TravelAgency.model;

public class Reservation : Entity<int>
{
    private Trip _trip;
    private String _clientName;
    private String _clientContact;
    private int _numberOfReservedSeats;

    public Reservation(Trip trip, String clientName, String clientContact, int numberOfReservedSeats)
    {
        _trip = trip;
        _clientName = clientName;
        _clientContact = clientContact;
        _numberOfReservedSeats = numberOfReservedSeats;
    }

    public Trip Trip => _trip;

    public String ClientName
    {
        get => _clientName;
        set => _clientName = value;
    }

    public String ClientContact
    {
        get => _clientContact;
        set => _clientContact = value;
    }

    public int NumberOfReservedSeats
    {
        get => _numberOfReservedSeats;
        set => _numberOfReservedSeats = value;
    }

    public override string ToString()
    {
        return _trip.ToString() + $"Client Name: {_clientName}, Client Contact: {_clientContact}, Number of Reserved Seats: {_numberOfReservedSeats}";
    }

    public override bool Equals(object obj)
    {
        if (obj == null || GetType() != obj.GetType())
            return false;
        Reservation other = (Reservation)obj;
        return _trip == other._trip && _clientContact == other._clientContact && _clientName == other._clientName && 
               _numberOfReservedSeats == other._numberOfReservedSeats;

    }

    public override int GetHashCode()
    {
        return HashCode.Combine(_trip, _clientName, _clientContact, _numberOfReservedSeats);
    }
}

File: Fourth Semester - Design and Programming Environments/Java/ProiectJava/java/org/example/modul/Reservation.java
package org.example.modul;

import java.util.Objects;

public class Reservation extends Entity<Integer> {
    private Trip travel;
    private String clientName;
    private String clientContact;
    private Integer numberOfReservedSeats;

    public Reservation(Trip travel, String clientName, String clientContact, Integer numberOfReservedSeats) {
        this.travel = travel;
        this.clientName = clientName;
        this.clientContact = clientContact;
        this.numberOfReservedSeats = numberOfReservedSeats;
    }

    public Trip getTrip() {
        return travel;
    }

    public void setTrip(Trip travel) {
        this.travel = travel;
    }

    public String getClientName() {
        return clientName;
    }

    public void setClientName(String clientName) {
        this.clientName = clientName;
    }

    public String getClientContact() {
        return clientContact;
    }

    public void setClientContact(String clientContact) {
        this.clientContact = clientContact;
    }

    public Integer getNumberOfReservedSeats() {
        return numberOfReservedSeats;
    }

    public void setNumberOfReservedSeats(Integer numberOfReservedSeats) { this.numberOfReservedSeats = numberOfReservedSeats; }


    @Override
    public String toString() {
        return "Reservation{" +
                "travel=" + travel +
                ", nameClient='" + clientName +
                ", contactClient='" + clientContact +
                ", numberOfReservedPlaces=" + numberOfReservedSeats +
                '}';
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if(!(obj instanceof Reservation reservation)) {
            return false;
        }
        return getTrip().equals(reservation.getTrip()) && getClientName().equals(reservation.getClientName()) &&
                getClientContact().equals(reservation.getClientContact()) &&
                getNumberOfReservedSeats().equals(reservation.getNumberOfReservedSeats());
    }

    @Override
    public int hashCode() {
        return Objects.hash(getTrip(), getClientName(), getClientContact(), getNumberOfReservedSeats());
    }
}
